### DISCUSSION TITLE
Access and modify pipeline's widget visibility status

### DISCUSSION THREAD

**Discussion by JonathanCapuana** [Kitware/trame#440] [General]
![image](https://github.com/Kitware/trame/assets/74074836/62f678eb-c069-4b2b-a488-1c748f8d93a0)
Good day,

I am working on a project where I need to interact programmatically with the state of pipeline widgets. Specifically, I'd like to check their current status and use this information for conditional statements, as well as to be able to modify these states.

For example, as seen in the attached image, I am looking to activate the widget's state when the 'Apply' button for creating streamlines is pressed. Essentially, I want the widget to reflect that it has been activated or is 'in use' once the 'Apply' button is engaged.

Is there an existing method or property that allows accessing and modifying the widget states within the pipeline? Any guidance or code examples would be greatly appreciated.

Thank you for your assistance.

**Comment by jourdain** [comment]
It seems that you just need to update the pipeline state. It is like any other state variable so I'm not sure what example you actually need.

**Reply by JonathanCapuana** [reply]
Thank you for your quick reply. I apologize if my initial query wasn't clear enough.

What I'm looking to achieve is a way to monitor the visibility state of the streamlines within my application. I understand that it might be a simple state variable update, but I'm unsure how to implement this in practice.

To clarify, I need a variable that automatically updates to "1" when, for example, the streamlines are made visible and changes to "0" when they are hidden. This is so I can use this variable as a flag for conditional logic in other parts of my script.

Moreover, I would like to know how to change the visibility state of the streamlines programmatically from my script, not just through the user interface. 

I hope this clarifies my request, and I appreciate your help

**Reply by jourdain** [reply]
You may want to go over the trame tutorial. That may help you understanding how to do what you aim to do.

**Comment by JonathanCapuana** [comment]
Hi Jourdain
As you suggested, I thoroughly explored the tutorials and found discussions relevant to my problem (243, 221, 115, and 266), but I couldn't find a solution.

Here is my definition of the pipeline and a function that should change the visibility of the mesh element and update the state, created just as example:

```python
def switch_visibility():
    mesh_actor.SetVisibility(0)
    for node in state.pipeline: 
        if node["id"] == "1":
            node["visible"] = 0
    state.dirty("pipeline")
    ctrl.view_update()

def pipeline_widget():
    trame.GitTree(
        sources=(
            "pipeline",
            [
                {"id": "1", "parent": "0", "visible": 1, "name": "Mesh"},
                {"id": "2", "parent": "1", "visible": 1, "name": "Contour"},
            ],
        ),
        actives_change=(actives_change, "[$event]"),
        visibility_change=(visibility_change, "[$event]"),
    )
```

   In order to see the updated status on the pipeline i need to refresh the page.
    Thank you in advance, Jonathan

**Comment by jourdain** [comment]
The following code is working for me

```python
import asyncio
from trame.app import get_server
from trame.ui.html import DivLayout
from trame.widgets import trame
from trame.decorators import TrameApp, controller


def toggle_node(node, node_id):
    if node.get("id") == node_id:
        new_visible = (node.get("visible") + 1) % 2
        return {**node, "visible": new_visible}
    return node


@TrameApp()
class App:
    def __init__(self, server=None):
        self.server = get_server(server, client_type="vue3")
        self.ui = self._build_ui()

    @property
    def state(self):
        return self.server.state

    def toggle_visibility(self, node_id):
        with self.state:
            self.state.pipeline = [toggle_node(n, node_id) for n in self.state.pipeline]

    @controller.add_task("on_server_ready")
    async def auto_toggle(self, **kwargs):
        while True:
            for node_id in ["1", "2"]:
                await asyncio.sleep(0.5)
                self.toggle_visibility(node_id)

    def _build_ui(self):
        with DivLayout(self.server) as layout:
            trame.GitTree(
                sources=(
                    "pipeline",
                    [
                        {"id": "1", "parent": "0", "visible": 1, "name": "Mesh"},
                        {"id": "2", "parent": "1", "visible": 1, "name": "Contour"},
                    ],
                ),
            )


if __name__ == "__main__":
    app = App()
    app.server.start()
```
